packages/ucef/contracts/UCEFRegulated.sol

```diff
@@ -124,7 +133,111 @@ contract UCEFRegulated is UCEF {
      * - Both account owners and the regulator can view balances
      */
     function _authorizeBalance(address account) internal view virtual override returns (bool) {
-        require(msg.sender == account || msg.sender == _regulator, "Unauthorized access to balance");
+        if (msg.sender != account && msg.sender != _regulator) {
+            revert UCEFUnauthorizedBalanceAccess(msg.sender, account);
+        }
         return true;
     }
+
+    /**
+     * @dev Override to implement regulated event visibility for transfers
+     * Regulator can view all transfers for oversight alongside the involved parties
+     * @param from The sending address
+     * @param to The receiving address
+     * @return allowedViewers Array containing sender, receiver, and regulator
+     */
+    function _getTransferEventViewers(
+        address from,
+        address to,
+        uint256 /* value */
+    ) internal view virtual override returns (address[] memory allowedViewers) {
+        // Count unique non-zero addresses including regulator
+        uint256 viewerCount = 1; // Always include regulator
+        if (from != address(0) && from != _regulator) viewerCount++;
+        if (to != address(0) && to != from && to != _regulator) viewerCount++;
+
+        allowedViewers = new address[](viewerCount);
+        uint256 index = 0;
+
+        // Populate viewer list
+        allowedViewers[index++] = _regulator; // Always include regulator
```

Comment:
It's good to note for this case and any case of contracts with "read superusers" (I'll call them "superreaders"), that if the "superreader" set changes (like the regulator is changed, or new readers are added/removed), the specific historic event will always remain readable for the "at the time superreaders" (I think that is fine, they were able to read the event at the time anyway), but the new superreaders won't have access to it. Might be sth to consider in the design phase, but probably not at the contract level, more like on the SD:R RPC level. In that sense, maybe RPC should be reading a predefined function like `_getTransferEventViewers`, to always get the up-to-date superreader list. CC: @campbell-codes